### PR TITLE
Opt out telemetry command

### PR DESCRIPTION
**Pull request by ivanThePleasant**
### What does it do?

This PR introduces a CLI command that allows users to opt out of Strapi telemetry without the need to delete their project `uuid` from `package.json` file or provide additional `env` variables. CLI command adds `"optOutTelemetry": true` line to the project's `package.json` file and notifies analytics that telemetry has been disabled.

### Why is it needed?

Currently disabling telemetry is not straightforward and there is also no way to tell how many projects opted out so that we could build a bigger picture. Also, deleting project `uuid` from `package.json` might not be a good idea since some unique project identifier might be necessary for the project in the future, but adding it back to `package.json` will enable the telemetry again. Proposed approach is cleaner and more functional.

### How to test it?

You can test this command directly in the get started strapi project in the monorepo. Run `yarn strapi opt-out-telemetry` from the root of `examples/getstarted`.

### CODE REVIEW

File: packages/core/strapi/bin/strapi.js
Comment (ivanThePleasant) [reply packages/core/strapi/bin/strapi.js:233]: Thanks, good catch! 😊

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (fdel-car) [inline packages/core/strapi/lib/commands/opt-out-telemetry.js:48]: The result of the `resolve` function will always return something (even if the package.json doesn't exist) since it just processes path arguments.

So what you are trying to do here won't work, this condition will if, I'm not mistaken, never be run.

By the way, you can remove the `process.cwd()` call just above since you didn't prefix `package.json` with a `/`, `resolve` will try to resolve the final path from the current working directory by default.

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (fdel-car) [inline packages/core/strapi/lib/commands/opt-out-telemetry.js:64]: Maybe it's not important but should we call this property `telemetryDisabled` to keep it in line with the existing environment variable naming? It's also a bit more straightforward naming I feel.

Not asking to rename the event name we send, just this `package.json` property :)

**Comment by fdel-car** [comment]
Just wanted to say that I like this idea of a CLI command to opt-out from the telemetry, it's less destructive than removing the UUID and more straightforward as you said 👍

**Review by fdel-car — CHANGES_REQUESTED** [review]

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:48]: I chose `process.cwd()` here because for me it just looks a bit more readable in accordance with their documentations, that all.
Regarding the check for packageJSON - it's there only to check if for some reason the `resolve()` itself malfunctions. But there are other things up the chain which check if command is run from a strapi project directory and if package.json is in place. Just a precaution against `resolve()` malfunctioning for some reason and returning nothing.

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js]: I like these suggestions, thank you. Have made changes 👍

**Review by fdel-car — DISMISSED** [review]

**Comment by ivanThePleasant** [comment]
I'll wait for at least one more approval before merging

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (gu-stav) [inline packages/core/strapi/lib/commands/opt-out-telemetry.js:74]:
```suggestion
      )}: There has been an error, please set "telemetryDisabled": true in the "strapi" object of your package.json manually.`
```

Typo :)

**Review by gu-stav — CHANGES_REQUESTED** [review]
One last typo: rest looks very good ✨

**Review by gu-stav — DISMISSED** [review]
✨ 🚀

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/commands/opt-out-telemetry.js:12]: You will have to use lodash.get this could be break if the user decided to remove the strapi key

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/commands/opt-out-telemetry.js:14]: since uuid is in the package why do you return two separate values? the function name is a bit confusing I think

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/commands/opt-out-telemetry.js:46]:
```suggestion
  const packageJSONPath = resolve(process.cwd(), 'package.json');
```

Comment (alexandrebodin) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:46]: it's misleading to have packageJSON be a path and not the value itself

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
The idea is there 👍 , just some clarification to add to the code

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:12]: I have implemented the check for `uuid` in the main function so I would not need to bring in additional packages. The process will be terminated if `uuid` will be undefined since that automatically means user already opted out. Is there a flow in this logic and I would still need to use lodash?

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:14]: Habit of separate vars for everything 😅 No real reason, just nicer for my subjective eye

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:46]: Agree, this is misleading. Will change, thanks 👍

File: packages/core/strapi/bin/strapi.js
Comment (ivanThePleasant) [reply packages/core/strapi/bin/strapi.js:232]: I was thinking of creating a separate PR for enabling it to limit the amount of files. You think it's better idea to keep it a part of one PR instead?
Sure, `telemetry:disable` and `telemetry:enable` will sound nice

Comment (alexandrebodin) [reply packages/core/strapi/bin/strapi.js:232]: Sure you can only do the telemetry:disable in this one no pbl

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (alexandrebodin) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:14]: we usually use object instead of arrays in the codebase for this then :)

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/commands/opt-out-telemetry.js:48]: this doesn't check if the package.json file exists this is just a path here. you need to check with fs-extra and `pathExists`

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (alexandrebodin) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:12]: Hum I don't see where you check if packageJson.strapi exists.

You also can have an error on L55 with `packageObj.strapi.telemetryDisabled` 

if the strapi key is not in the package.json all of this will breakfrom what I'm seeing. have you tried it?

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (alexandrebodin) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:12]: Here is the exact error you get 

```
Error: Cannot read property 'uuid' of undefined
TypeError: (intermediate value) is not iterable
    at optOutTelemetry (/Users/alex/dev/strapi/strapi-v4/packages/core/strapi/lib/commands/opt-out-telemetry.js:53:30)
```

Not very friendly :)

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:14]: Object it is

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:48]: There is no such check in this file. Instead, the command is invoked from `packages/core/strapi/bin/strapi.js` and there is this check implemented, which prevents the command from running if no `package.json` is present, or if `package.json` is not a strapi one:
```javascript
const checkCwdIsStrapiApp = name => {
  let logErrorAndExit = () => {
    console.log(
      `You need to run ${yellow(
        `strapi ${name}`
      )} in a Strapi project. Make sure you are in the right directory.`
    );
    process.exit(1);
  };

  try {
    const pkgJSON = require(process.cwd() + '/package.json');
    if (!_.has(pkgJSON, 'dependencies.@strapi/strapi')) {
      logErrorAndExit(name);
    }
  } catch (err) {
    logErrorAndExit(name);
  }
};

Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:48]: 👍

Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js:48]: But wouldn't yarn/npm throw an error regardless if `package.json` was not in place? It's not going to run anything since `strapi` command is also going to be unavailable, isn't it?

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (alexandrebodin) [inline packages/core/strapi/lib/commands/opt-out-telemetry.js:49]: You want to throw only if it doesn't exist no?

```suggestion
  if (!exists) {
```

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
a simple typo that break everything ;)

File: packages/core/strapi/lib/commands/opt-out-telemetry.js
Comment (ivanThePleasant) [reply packages/core/strapi/lib/commands/opt-out-telemetry.js]: You got me dazzled here. I was looking at the code for 10 minutes not understanding where did the "!" go. Must have pressed cmd+z one time too much before committing 🤦‍♂️

**Review by alexandrebodin — DISMISSED** [review]
I think it's missing a way to disable the telemetry on the front side at the moment. @soupette send a null uuid in the init route would do the trick ?

**Comment by ivanThePleasant** [comment]
I'm working on that, will push update shortly

**Comment by soupette** [comment]
> I think it's missing a way to disable the telemetry on the front side at the moment. @soupette send a null uuid in the init route would do the trick ?

Yes that's it the `uuid` is returned by the `/admin/init` endpoint so if it returns null it disabled for the entire admin panel.

**Comment by ivanThePleasant** [comment]
I have added necessary changes so that front side events also take into account `telemetryDisabled` flag from the `package.json`. Deleting `uuid` functionality in order to disable the telemetry is also preserved. Had to slightly change sequence of events in couple of places.

**Comment by ivanThePleasant** [comment]
@alexandrebodin can the front-end tests be failing due to this https://strapihq.slack.com/archives/C02KEL5UBPZ/p1648575334515669 ?

**Comment by alexandrebodin** [comment]
@ivanThePleasant do we have a doc PR ready for this ?

**Comment by ivanThePleasant** [comment]
@alexandrebodin I have already spoken to the documentation team, Shaun is waiting for me to give him the last version of what needs to be added to the docs. If this PR now has green light, then I will update Shaun with all the changes and we'll prepare the PR asap

**Review by soupette — APPROVED** [review]

**Comment by ivanThePleasant** [comment]
Docs team has been notified about the changes and are going to prepare a PR with documentation changes. Please don't merge this PR until then. Docs PR should be ready by next week (commencing 4th April).

**Review by fdel-car — APPROVED** [review]

**Pull request merged**
